### PR TITLE
Add fix for multiple_activations test

### PR DESCRIPTION
**Pull request by kbruniec**
- Add forbid activation fusing pass for GNA2 library
   - Fix get all prev layers fnct
   - To enable GNA_NoRegression.smoke_multiple_activations_onGNA_INT16

### CODE REVIEW

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 1]
```diff
@@ -268,6 +269,40 @@ void HandleMultipleActivationsForTheLayerPass::run() {
...
+        if (LayerInfo(l).isActivation()) {
+            if (LayerInfo(prevLayer).has32BOutput()) {
+                // find all layers directly connected to the outputs of the previous layer
+                const auto allUsingPrev = CNNNetGetAllNextLayersSkipCertain(prevLayer, -1,
+                    [&](CNNLayerPtr nextLayer) -> bool {
+                        for (const auto& input : nextLayer->insData) {
+                            for (const auto& output : prevLayer->outData) {
+                                if (areEqualDatas(input.lock(), output) &&
+                                    areEqualDatas(l->insData[0].lock(), output) &&
+                                    ((nextLayer->type == "Eltwise" && nextLayer->params.at("operation") == "sum")
```
Comment (bsochack) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:287]: it is better to use LayerInfo and isEltwiseSum()

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 2]
```diff
@@ -268,6 +269,40 @@ void HandleMultipleActivationsForTheLayerPass::run() {
     }
 }
 
+void ForbidActivationFusingPass::run() {
+    for (auto& l : *pLayers) {
+        if (l->insData.empty()) continue;
+        auto prevLayer = CNNNetPrevLayerSkipCertain(l, 0, [](CNNLayerPtr ptr) {
+            return LayerInfo(ptr).isNonFunctional();
+            });
+        if (LayerInfo(l).isActivation()) {
```
Comment (bsochack) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:278]: I guess it is enough to detect if it is activation layer.
Then for activation layer check if previous layer has more than 1 layer connected. If it is the case, then the diagonal layer needs to be added.
I don't think you need to skip any layers when you search for all directly connected layers to the previous one. If you have other opinion, please add in comment what cases need to be skipped and why.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.hpp [hunk 3]
```diff
@@ -123,6 +123,14 @@ DECL_PASS(ReorderMaxPool);
...
+ * @brief GNA doesn't provide intermediate results (sums) when the layer is fused with activation.
+ * When more layers use the sums as inputs (beside the activation) then the diagonal layer
+ * is inserted before the activation to forbid the fusing and make the sums exposed.
+ * This is observed in the multiple_activations_onGNA_INT16 test.
+ */
+DECL_PASS(ForbidActivationFusing);
+
 /**
  * @brief copy layer insertion required in cases where input layer does not have output memory
  */
```
Comment (bsochack) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.hpp:136]: Please add tests or unblock some if there are already existing ones for this issue.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp [hunk 4]
```diff
@@ -268,6 +269,40 @@ void HandleMultipleActivationsForTheLayerPass::run() {
     }
 }
 
+void ForbidActivationFusingPass::run() {
```
Comment (bsochack) [inline inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:272]: This pass could be implemented as part of InsertDiagonalLayerPass

File: inference-engine/src/gna_plugin/gna_graph_tools.hpp [hunk 5]
```diff
@@ -220,9 +220,14 @@ inline std::pair<InferenceEngine::CNNLayerPtr, int>  CNNNetCheckNextLayerSkipCer
             separate_layers(getInputTo(layer->outData[i]));
         }
 
+        std::set< CNNLayerPtr > visited;
```
Comment (bsochack) [inline inference-engine/src/gna_plugin/gna_graph_tools.hpp:223]: If you change a type of container for currentSet, you will not have to deal with duplicates.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.hpp (hunk 3 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.hpp:136]: The test is in different/internal repo: git@gitlab-icv.inn.intel.com:inference-engine/dldt-tests-closed.git

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 4 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:272]: It was done like that in the first approach. Then I have added new pass to make the code separated for better maintainability and not complicate InsertDiagonalLayerPass() more.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 1 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:287]: Thanks, I will try with that

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 2 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:278]: It is done that way. But I can try to simplify this by reordering and using CNNNetPrevLayer() instead CNNNetPrevLayerSkipCertain().

File: inference-engine/src/gna_plugin/gna_graph_tools.hpp (hunk 5 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/gna_graph_tools.hpp:223]: I don't see it. IMO it is not the problem with duplicates, but circular dependency.
Please see that element is inserted to `currentSet` in `separate_layers` lambda.
So after removal of the element the same element can be inserted once again (we have to track which elements were visited).

Comment (bsochack) [reply inference-engine/src/gna_plugin/gna_graph_tools.hpp:223]: Ok, it makes sense.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 1 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp]: It's done.

File: inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp (hunk 2 above)
Comment (kbruniec) [reply inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp:274]: It's simplified now.

**Review by bsochack — APPROVED** [review]

**Pull request merged**